lib/Runtime/Library/ByteCodeString.h

```diff
@@ -0,0 +1,21 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+#pragma once
+
+namespace Js
+{
+    class ByteCodeString : public JavascriptString
```

Comment:
If I'm reading this right, ByteCodeString is a pointer to a string plus a pointer to the thing that owns that string data, so the string is guaranteed to stay alive. This sounds a lot like SubString. Can we just use SubString?

Reply:
If we use SubString, we also get the benefit of pre-existing code that can turn it into a LiteralStringWithPropertyStringPtr when appropriate.

---
In reply to: [198550710](https://github.com/Microsoft/ChakraCore/pull/5374#discussion_r198550710) [](ancestors = 198550710)

Reply:
I guess the risk with using SubString is that we might unnecessarily copy the data when somebody calls GetSz because we didn't know it was null-terminated. (Side note: GetString plus GetLength is almost always the right thing to do rather than GetSz because JS allows embedded nulls, but a whole lot of existing code calls GetSz so we have to deal with it.) Maybe the cleanest option would be a new subclass of SubString that happens to know its data is null-terminated (and asserts that it is during construction). This NullTerminatedSubString would have to override GetSz to call JavascriptString::GetSz (bypassing the SubString::GetSz implementation) to avoid the copy. It would also need its own implementation of CloneToScriptContext because currently we rely on GetSz copying a SubString's buffer when marshaling. I'm not sure actually whether the best behavior is to clone the string data when marshaling or make a new NullTerminatedSubString in the other context; I could imagine problems with either.

---
In reply to: [198555571](https://github.com/Microsoft/ChakraCore/pull/5374#discussion_r198555571) [](ancestors = 198555571,198553710,198550710)

Reply:
Looks like this is only used in generating stack traces, which is not a super frequent task, so I'd say just use SubString and avoid the complication of worrying about whether it copies during GetSz.

---
In reply to: [198562070](https://github.com/Microsoft/ChakraCore/pull/5374#discussion_r198562070) [](ancestors = 198562070,198555571,198553710,198550710)

Reply:
Doesn't a `SubString` have to be a pointer into an existing `JavascriptString`? My understanding of this class is it is to represent a new kind of string which lives in memory that isn't owned by the GC or by `chakra[core].dll`, since that combination is causing assertion failures. The bytecode memory is host-owned but they promise not to free it while it is in use, I think.

Reply:
SubString needs a pointer to some GC-allocated owning thing, but it doesn't need to be a JavascriptString. I was assuming Utf8SourceInfo is GC-allocated.

---
In reply to: [198567219](https://github.com/Microsoft/ChakraCore/pull/5374#discussion_r198567219) [](ancestors = 198567219)